test/browser/component.js

```diff
@@ -686,3 +686,68 @@ context(`$hooks`, function () {
     });
   });
 });
+
+describe(`Lifecycle Helpers`, function () {
+  let el;
+  let spy;
+
+  beforeEach(function () {
+    document.body.innerHTML = ``;
+    el = document.createElement(`simple-app`);
+    spy = sinon.spy();
+  });
+
+  it(`onConnected doesn't fire until connectedCallback`, async function () {
+    el.onConnected(spy);
+
+    expect(spy.callCount).to.equal(0);
+
+    document.body.appendChild(el);
+    await nextAnimationFrame();
+    expect(spy.callCount).to.equal(1);
+  });
+
+  it(`onConnected immediately fires if connectedCallback has run`, async function () {
+    document.body.appendChild(el);
+    await nextAnimationFrame();
+
+    el.onConnected(spy);
+    expect(spy.callCount).to.equal(1);
+  });
+
+  it(`onDisconnect fires on disconnectedCallback`, async function () {
+    document.body.appendChild(el);
+    await nextAnimationFrame();
+
+    el.onDisconnected(spy);
+    expect(spy.callCount).to.equal(0);
+
+    document.body.innerHTML = ``;
+    await nextAnimationFrame();
+    expect(spy.callCount).to.equal(1);
+  });
+
+  it(`onConnect can return a function to run on disconnectedCallback`, async function () {
+    document.body.appendChild(el);
+    await nextAnimationFrame();
+
+    el.onConnected(() => spy);
+    expect(spy.callCount).to.equal(0);
+
+    document.body.innerHTML = ``;
+    await nextAnimationFrame();
+    expect(spy.callCount).to.equal(1);
+  });
+
+  it(`onConnect function context is bound to the component`, async function () {
+    document.body.appendChild(el);
+    await nextAnimationFrame();
+
+    el.onConnected(function () {
+      this.update({baz: 42});
+    });
+    await nextAnimationFrame();
+
+    expect(el.textContent).to.contain(`baz: 42`);
+  });
```

Comment:
to my point above about components getting moved around in the DOM, we'll want test cases in here where the component gets taken out of the DOM and inserted somewhere else (or the same place, doesn't matter, it'll still cause the Custom Elements callbacks to fire).

Once those are working we might want a couple isorender test cases